staging/src/k8s.io/kube-aggregator/pkg/apiserver/openapi_aggregator.go

```diff
@@ -36,47 +36,98 @@ import (
+
 	// provided for dynamic OpenAPI spec
 	openAPIService *handler.OpenAPIService
-
-	// Aggregator's OpenAPI spec (holds apiregistration group).
-	aggregatorOpenAPISpec *spec.Swagger
-
-	// Local (in process) delegate's OpenAPI spec.
-	inProcessDelegatesOpenAPISpec *spec.Swagger
-
-	contextMapper request.RequestContextMapper
 }
 
-func buildAndRegisterOpenAPIAggregator(delegateHandler http.Handler, webServices []*restful.WebService, config *common.Config, pathHandler common.PathHandler, contextMapper request.RequestContextMapper) (s *openAPIAggregator, err error) {
-	s = &openAPIAggregator{
-		openAPISpecs:  map[string]*openAPISpecInfo{},
-		contextMapper: contextMapper,
+var _ OpenAPIAggregationManager = &openAPIAggregator{}
+
+// This function is not thread safe as it only being called on startup.
+func (s *openAPIAggregator) addLocalSpec(spec *spec.Swagger, localHandler http.Handler, name, etag string) {
+	localApiService := apiregistration.APIService{}
+	localApiService.Name = name
+	s.openAPISpecs[name] = &openAPISpecInfo{
+		etag:       etag,
+		apiService: localApiService,
+		handler:    localHandler,
+		spec:       spec,
 	}
+}
 
-	// Get Local delegate's Spec
-	s.inProcessDelegatesOpenAPISpec, err = s.downloadOpenAPISpec(delegateHandler)
-	if err != nil {
-		return nil, err
+// This function is not thread safe as it only being called on startup.
+func buildAndRegisterOpenAPIAggregator(downloader *openAPIDownloader, delegationTarget server.DelegationTarget, webServices []*restful.WebService,
+	config *common.Config, pathHandler common.PathHandler) (s *openAPIAggregator, err error) {
+	s = &openAPIAggregator{
+		openAPISpecs: map[string]*openAPISpecInfo{},
 	}
 
+	i := 0
 	// Build Aggregator's spec
-	s.aggregatorOpenAPISpec, err = builder.BuildOpenAPISpec(
+	aggregatorOpenAPISpec, err := builder.BuildOpenAPISpec(
 		webServices, config)
 	if err != nil {
 		return nil, err
 	}
-	// Remove any non-API endpoints from aggregator's spec. aggregatorOpenAPISpec
-	// is the source of truth for all non-api endpoints.
-	aggregator.FilterSpecByPaths(s.aggregatorOpenAPISpec, []string{"/apis/"})
+	aggregator.FilterSpecByPaths(aggregatorOpenAPISpec, []string{"/apis/"})
+
+	// Reserving non-name spec for aggregator's Spec.
+	s.addLocalSpec(aggregatorOpenAPISpec, nil, fmt.Sprintf(localDelegateChainNamePattern, i), "")
+	i++
+	for delegate := delegationTarget; delegate != nil; delegate = delegate.NextDelegate() {
+		handler := delegate.UnprotectedHandler()
+		if handler == nil {
+			continue
+		}
+		delegateSpec, etag, _, err := downloader.Download(handler, "")
+		if err != nil {
+			return nil, err
+		}
+		if delegateSpec == nil {
+			continue
+		}
+		specName := fmt.Sprintf(localDelegateChainNamePattern, i)
+
+		// We would find first spec with "/api/" endpoints and consider it the
+		// base for the merge. Any other spec would be filtered for their "/apis/"
+		// endpoints only.
+		filterPaths := true
+		if len(s.baseForMergeSpecName) == 0 {
```

Comment:
do we check that we eventually find one? On the other hand, having not `/api` is a completely normal case. Then we probably should choose the first delegate. Or maybe we should just mark this one (with a bool in the spec info that this provides the /api? Then we can filter accordingly.

Reply:
Not even sure that we do need this bool. If we filter out the `/api` paths from all other specs than the first having `/api`, we are fine, aren't we? Then order in `buildOpenAPISpec` for the local services does not matter anymore. Wdyt?

Reply:
Or can this even happen that two delegates have `/api`? It shouldn't. If it does it's a conflict that should be fatal for the aggregation.

Reply:
Hmm, if the case with no `/api` is a valid case, then we can simply point `baseForMergeSpecName` to first delegate. The question is should it be aggregation server or the one after? In normal kubernetes case, is the one after. I thought about adding a boolean, but then I need to make sure one of APIServers will have that boolean. It is just scream bug :) having `baseForMergeSpecName` will make sure we would not have two server having the boolean set. I think we can proceed with setting aggregation API server as the default one if no one provides `/api`. We do get other non-api information from this server (like Swagger.Info, ...) so we need to have one.

Reply:
Does anything speak against using the aggregator as base all the time? Does it have all "non-api information" ?